arch/risc-v/src/rv32im/Toolchain.defs

```diff
@@ -67,13 +69,6 @@ ifeq ($(CONFIG_RISCV_TOOLCHAIN),GNU_RVG)
   endif
 endif
 
-# Individual tools may limit the optimization level but, by default, the
-# optimization level will be set to -Os
-
-ifeq ($(CONFIG_DEBUG_SYMBOLS),)
-MAXOPTIMIZATION ?= -Os
```

Comment:
should we unify the behaviour for all boards?

Reply:
Do you for all RISC-V boards? Or as a top-level definition in the NuttX build system?
From what I verified, at least for RV32IM-based boards is somewhat unified. Each board is able to override the `MAXOPTIMIZATION` at the bottom-level Make.defs.

Reply:
I mean all board not onldy risc-v.

Reply:
We need to understand in the first place why it was defined this way. I think @patacongo did this way to force small binaries size when DEBUG Symbols is not used. In the other hand when DEBUG Symbols is enabled  the "-Os" will mess with the GDB source file/line association.

Reply:
I always wondered if using -Og wouldn't be better for debugging with optimizations enabled:

```
-Og

    Optimize debugging experience. -Og should be the optimization level of choice for the standard edit-compile-debug cycle, offering a reasonable level of optimization while maintaining fast compilation and a good debugging experience. It is a better choice than -O0 for producing debuggable code because some compiler passes that collect debug information are disabled at -O0.

    Like -O0, -Og completely disables a number of optimization passes so that individual options controlling them have no effect. Otherwise -Og enables all -O1 optimization flags except for those that may interfere with debugging:


```

Reply:
> We need to understand in the first place why it was defined this way. I think @patacongo did this way to force small binaries size when DEBUG Symbols is not used. In the other hand when DEBUG Symbols is enabled the "-Os" will mess with the GDB source file/line association.

Although I understand if this is indeed the original motivation, this makes the build system inconsistent. when the user selects `CONFIG_DEBUG_SYMBOLS` the expected output is the exact same code, plus debug symbols. The build system should not infer other scenarios.



> I always wondered if using -Og wouldn't be better for debugging with optimizations enabled:
> 
> ```
> -Og
> 
>     Optimize debugging experience. -Og should be the optimization level of choice for the standard edit-compile-debug cycle, offering a reasonable level of optimization while maintaining fast compilation and a good debugging experience. It is a better choice than -O0 for producing debuggable code because some compiler passes that collect debug information are disabled at -O0.
> 
>     Like -O0, -Og completely disables a number of optimization passes so that individual options controlling them have no effect. Otherwise -Og enables all -O1 optimization flags except for those that may interfere with debugging:
> ```

I'd say `-Og` is a good alternative to `-O0`, but again, the build system should not change the optimization level deliberately.

Reply:
> > We need to understand in the first place why it was defined this way. I think @patacongo did this way to force small binaries size when DEBUG Symbols is not used. In the other hand when DEBUG Symbols is enabled the "-Os" will mess with the GDB source file/line association.
> 
> Although I understand if this is indeed the original motivation, this makes the build system inconsistent. When the user selects `CONFIG_DEBUG_SYMBOLS` the expected output is the exact same code, plus debug symbols. The build system should not infer other scenarios.
> 
Yes, but imagine the developer wants to debug this code in the GDB and enable the DEBUG Symbols and everything get messed, he/she will complain at NuttX mailing list. This modification is a fallback to avoid it.

Sure, if -Og works correctly it could be a good idea. I don't think is it a good idea to apply this modification globally without testing it on other MCUs.

Reply:
> > > We need to understand in the first place why it was defined this way. I think @patacongo did this way to force small binaries size when DEBUG Symbols is not used. In the other hand when DEBUG Symbols is enabled the "-Os" will mess with the GDB source file/line association.
> > 
> > 
> > Although I understand if this is indeed the original motivation, this makes the build system inconsistent. When the user selects `CONFIG_DEBUG_SYMBOLS` the expected output is the exact same code, plus debug symbols. The build system should not infer other scenarios.
> 
> Yes, but imagine the developer wants to debug this code in the GDB and enable the DEBUG Symbols and everything get messed, he/she will complain at NuttX mailing list. This modification is a fallback to avoid it.

The user will be debugging a different code, one that might even behave differently and not present any issue at all. Or might not even fit the flash of the target device. This decision hurts the consistency of the build system.

The correct solution is to instruct the user to reduce the optimization level (e.g. `-Og`) and try again. It is not the OS responsibility to infer these details. The OS could offer this possibility as an option, but it needs to be explicit.

Reply:
Yes, I totally agree with @gustavonihei that -g option should be decoupled from -Oxxx option. -g option is very useful for the post-mortem debugging even with the highest optimization code. Actually, we always enable -g option in all build config. There isn't any harmful to enable -g all time if you account the storage capacity become bigger and bigger, but the price become cheeper and cheeper.